go/interpreter/lfvm/hash_cache_test.go

```diff
@@ -0,0 +1,330 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package lfvm
+
+import (
+	"fmt"
+	"math/rand"
+	"slices"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/Fantom-foundation/Tosca/go/tosca"
+)
+
+func TestSha3HashCache_hash_ProducesCorrectHashesForInputs(t *testing.T) {
+	inputs := [][]byte{
+		{},
+		{0},
+		{1, 2, 3, 4, 5},
+		make([]byte, 32),
+		make([]byte, 64),
+		make([]byte, 123),
+	}
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	for i := 0; i < 100; i++ {
+		input := make([]byte, r.Intn(150))
+		r.Read(input)
+		inputs = append(inputs, input)
+	}
+
+	cache := newSha3HashCache(10, 10)
+	for _, input := range inputs {
+		want := Keccak256(input)
+		got := cache.hash(input)
+		if want != got {
+			t.Errorf("expected hash to be %x, but got %x", want, got)
+		}
+	}
+}
+
+func TestHashCache_UsesProvidedHashingFunction(t *testing.T) {
+	hash := func(i int) tosca.Hash {
+		return tosca.Hash{byte(i)}
+	}
+
+	cache := newHashCache(10, hash)
+	for i := 0; i < 100; i++ {
+		want := hash(i)
+		got := cache.getHash(i)
+		if want != got {
+			t.Errorf("expected hash to be %x, but got %x", want, got)
+		}
+	}
+}
+
+func TestHashCache_HashesAreCached(t *testing.T) {
```

Comment:
shouldn't this test also check that counter increases ?

Reply:
The property that the provided hash function is actually used is covered by the function above. But to make it more self-contained, I added a line checking that the produced hashes are different.